### PR TITLE
Always simplify sympy expressions before printing.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127543

This is important because if a replacement has happened during inductor lowering, we may have stale symbols in sympy expressions that we need to replace away.  Do this at the very end.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -518,6 +518,11 @@ def _print_RoundDecimal(self, expr):
         assert isinstance(ndigits, sympy.Integer)
         return f"round({self._print(number)}, {ndigits})"
 
+    def doprint(self, expr, *, simplify: bool = True):
+        if simplify:
```
Comment (lezcano) [inline torch/_inductor/codegen/common.py:522]: Why is here better than before calling pexpr?

Comment (ezyang) [reply torch/_inductor/codegen/common.py:522]: There are a lot of pexpr calls so I figured I wanted to hit all of them

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -444,7 +444,7 @@ def __init__(self):
         self.stride = "stride()"
         self.last_seen_device_guard_index: Optional[int] = None
         self.supports_intermediate_hooks = True
-        self.expr_printer = pexpr
+        self.expr_printer: Callable[[Any], Any] = pexpr
```
Comment (Skylion007) [inline torch/_inductor/codegen/wrapper.py:447]: The return type here should still be string for this callable right?

**Comment by ezyang** [comment]
PR updated to also apply to C++ printer

**Review by lezcano — APPROVED** [review]
This is great. Feel free to merge once you convince CI.

**Pull request closed without merging**